composer/datasets/streaming/format.py

```diff
@@ -0,0 +1,227 @@
+from io import BufferedReader, BufferedWriter
+from typing import Dict, Sequence, Tuple
+
+import numpy as np
+
+
+def get_index_basename() -> str:
+    """Get the basename for a streaming dataset index.
+
+    Returns:
+        str: Basename of file.
+    """
+    return 'index.mds'
+
+
+def get_shard_basename(shard: int) -> str:
+    """Get the basename for a streaming dataset shard.
+
+    Args:
+        shard (int): Which shard.
+
+    Returns:
+        str: Basename of file.
+    """
+    return '%05d.mds' % shard
+
+
+def sample_dict_to_bytes(obj: Dict[str, bytes], keys: Sequence[str]) -> bytes:
+    """Dump a sample dict to bytes, given field names.
+
+    Args:
+        obj (dict): The sample dict to encode.
+        keys (list of str): The field names.
+
+    Returns:
+        bytes: The encoded sample bytes.
+    """
+    values = []
+    for key in keys:
+        value = obj[key]
+        values.append(value)
+    sizes = list(map(len, values))
+    sizes = np.array(sizes, np.int32)
+    return sizes.tobytes() + b''.join(values)
+
+
+def bytes_to_sample_dict(data: bytes, keys: Sequence[str]) -> Dict[str, bytes]:
+    """Load a sample dict from bytes and field names.
+
+    Args:
+        data (bytes): The encoded sample data.
+        keys (list of str): The field names.
+
+    Returns:
+        dict: The decoded sample dict.
+    """
+    num_values = len(keys)
+    sizes = np.frombuffer(data[:num_values * 4], np.int32)
+    ends = num_values * 4 + sizes.cumsum(0)
+    begins = ends - sizes
+    values = []
+    for begin, end in zip(begins, ends):
+        value = data[begin:end]
+        values.append(value)
+    return dict(zip(keys, values))
+
+
+class StreamingDatasetIndex(object):
+    """Streaming dataset index file, containing all the info about shards.
+
+    The shards are just dumb buffers with samples catted together. All the
+    offset info across the whole dataset is contained in the index file. Workers
+    read this file to calculate how much of which shards their slice is.
+
+    Each sample is a dict of str to bytes. All samples must contain the same
+    dict keys (fields). These strings are stored in the index file for efficiency.
+    """
+
+    def __init__(self, samples_per_shard: Sequence[int], bytes_per_shard: Sequence[int],
```

Comment:
Is it possible to use numpy arrays instead of python lists for these variables? Each python integer is 28 bytes minimum...a bit concerned about how much memory these lists will take up, especially for large datasets when replicated multiple times per node.

Reply:
Yeah, I was trying to be careful about that. In the Writer, where ram usage isn’t a concern, these values are accumulated in lists of int. Then in the Index, they are loaded from serialized format in np.ndarray of np.int32.